README.md
# pcnetm
```
pcnetm stands for path characteristics network measurement.
pcnetm is a rework of old pchar-1.5 to redesign and convert to modern C++.
```
[![C++](https://img.shields.io/badge/C%2B%2B-11-yellowgreen)](https://www.cppreference.com/) 
[![License](https://img.shields.io/badge/license-GPL3-_red.svg)](https://www.gnu.org/licenses/gpl-3.0.en.html)
### Obtaining old pchar

```
At present, the latest version of pchar is the 1.5 version released in 2005, which has not been updated for more than ten years. The software can be obtained by the following command:

wget http://www.kitchenlab.org/www/bmah/Software/pcnetm/pcnetm-1.5.tar.gz
```

### Description

       pcnetm  measures  the  characteristics  of  the network path between two Internet hosts, on
       either IPv4 or IPv6 networks.  It is  an  independently-written  reimplementation  of  the
       pathchar  utility, using similar algorithms.  Both programs measure network throughput and
       round-trip time by sending varying-sized UDP packets into the network and waiting for ICMP
       messages in response.  Like traceroute, they modulate the IPv4 time-to-live (TTL) field or
       the IPv6 hop limit field to get measurements at different distances along a path.

       In its default mode, a run of pcnetm over a short path might produce an output  that  looks
       like this:
       pcnetm to dancer.ca.sandia.gov (146.246.246.1) using UDP/IPv4
       Packet size increments by 32 to 1500
       46 test(s) per repetition
       32 repetition(s) per hop
        0:
           Partial loss:      0 / 1472 (0%)
           Partial char:      rtt = 0.657235 ms, (b = 0.000358 ms/B), r2 = 0.989713
                              stddev rtt = 0.004140, stddev b = 0.000006
           Hop char:          rtt = 0.657235 ms, bw = 22333.268771 Kbps
           Partial queueing:  avg = 0.000150 ms (418 bytes)
        1: 146.246.243.254 (con243.ca.sandia.gov)
           Partial loss:      0 / 1472 (0%)
           Partial char:      rtt = 0.811278 ms, (b = 0.000454 ms/B), r2 = 0.995401
                              stddev rtt = 0.003499, stddev b = 0.000005
           Hop char:          rtt = 0.154043 ms, bw = 83454.764777 Kbps
           Partial queueing:  avg = 0.000153 ms (336 bytes)
        2: 146.246.250.251 (slcon1.ca.sandia.gov)
           Partial loss:      0 / 1472 (0%)
           Partial char:      rtt = 1.044412 ms, (b = 0.002161 ms/B), r2 = 0.999658
                              stddev rtt = 0.004533, stddev b = 0.000006
           Hop char:          rtt = 0.233133 ms, bw = 4686.320952 Kbps
           Partial queueing:  avg = 0.000100 ms (46 bytes)
        3: 146.246.246.1 (dancer.ca.sandia.gov)
           Path length:       3 hops
           Path char:         rtt = 1.044412 ms, r2 = 0.999658
           Path bottleneck:   4686.320952 Kbps
           Path pipe:         611 bytes
           Path queueing:     average = 0.000100 ms (46 bytes)

       The  path  here  passes  through  three  hops.  Each hop consists of four lines of output:
       Partial loss documents the number and percentage of probe packets that  were  lost  during
       the  probes  for that hop.  The partial char line shows the estimated round-trip time from
       the probing host through the current hop.  The hop char line shows estimates of the round-
       trip  time  and  bandwidth  for  the  current hop.  Finally, the partial queueing shows an
       estimate of the average queueing along the path, up to and including the current hop.

       Between each hop, pcnetm prints the IP address and (if known) name of  the  host/router  at
       the end of the link.

       After  the last hop (usually the target host), pcnetm prints statistics on the entire path,
       including the path length and path pipe (the latter is an estimate of the  delay-bandwidth
       product of the path).

       pcnetm  has another mode of operation, called trout (short for “tiny traceroute”).  In this
       mode, packets of random sizes (one packet per hop diameter) are sent along the path  to  a
       destination.   No  attempt  at  estimating  link properties is made; however, this mode is
       extremely fast.  It is intended for use as a part of a larger measurement  infrastructure.
       The output from this mode might look like:
       trout to bmah-freebsd-1.cisco.com (171.70.84.44) using ICMP/IPv4 (raw sockets)
       Packet size increments from 28 to 1500 by 32
        0: 171.70.84.42 (bmah-freebsd-0.cisco.com)
        1: 171.70.84.44 (bmah-freebsd-1.cisco.com) 352 -> 352 bytes: 0.318 ms

### Usage

```
       -a analysis
              Set  analysis  type.   Current  choices are lsq (the default), which uses a minimum
              filter followed by a least sum-of-squares fit to estimate link bandwidths, kendall,
              which uses the same minimum filter followed by a linear fit based on Kendall's test
              statistic, lms, which does a minimum filter followed by a least median  of  squares
              fit,  and  lmsint,  which  is  an implementation of the lms computations using only
              integer arithmetic.

       -b burst
              Set the size of packet bursts.  A burst parameter > 1 will result in some number of
              ICMP_ECHOREPLY  packets  sent  before  the  probe packet to induce queueing.  These
              packets are useful for  measuring  store-and-forward  switched  subnets,  but  make
              measurements of fast links behind bottlenecks inaccurate.

       -c     Ignore  routing  changes  detected during running.  Normally, pcnetm will exit if it
              receives responses from more than one host for a  given  hop,  assuming  that  this
              condition  is  caused  by  a  routing  transient.   However, certain load-balancing
              schemes can also cause this condition.  In such situations, using the -c option may
              be useful.

       -C     Use pcap(3) packet capture library (this must have been enabled at configure time).
              Note that this option must be specified to enable TCP-based probes.

       -d debug
              Sets debugging output level.  Generally not useful except to the developer.

       -g gap Set the mean inter-probe gap in seconds.  The default is  0.25,  which  results  in
              approximately  four  probes  per  second  being  run.   Care should be taken not to
              decrease this gap by too much, in order to avoid flooding the network.  The default
              value  here  is  deliberately  conservative; users with the need or desire to probe
              more quickly are presumed to have  at  least  perused  the  documentation  for  the
              relevant command-line options.

       -G gaptype
              Set  distribution  used  to  select interprobe gap times.  Current alternatives are
              fixed  (the  default)  and  exp,  which  picks  gap  times  from   an   exponential
              distribution.   The  latter  option  is an attempt to simulate a Poisson process of
              probe packets (a lot of aliteration), however due  to  the  fact  that  each  probe
              experiment takes a non-zero amount of time, this is only an approximation.

       -H hops
              Set the maximum number of hops that pcnetm will probe into the network.  The default
              maximum is 30 hops, the same as with pathchar and traceroute.

       -h     Print usage information.

       -i interface
              Set the interface to listen on for the -C option.

       -I increment
              Set the probe packet size increment.  pcnetm will send IP packets  with  sizes  that
              are  integer  multiples of increment, up to the maximum specified by the -m option.
              The default is a 32-byte increment.  Small increments should produce more  accurate
              results, but will result in more probes (thus taking longer to run).

       -l origin
              Set the local source of probe packets.  This option is mostly useful on multi-homed
              hosts.  If not specified, it defaults to the value of hostname(3).  Note that  this
              option  must  be  used  if the local hostname cannot be resolved to an IPv4 or IPv6
              address.

       -m mtu Set the maximum probe packet size.  This value should be no larger  than  the  path
              MTU between the two hosts.  The default is 1500 bytes, the Ethernet MTU.

       -M mode
              Set  operational  mode.   The  normal  operational mode is pcnetm, which uses active
              probes to characterize the bandwidth, latency, loss,  and  queueing  of  the  links
              comprising  a path.  Another mode is trout, a “tiny traceroute” that is intended to
              be used as a portion of a larger network management infrastructure.

       -n     Don't attempt to resolve host addresses to names.

       -p protocol
              Select protocol to use.  Current options are: ipv4udp (UDP over IPv4), ipv4raw (UDP
              over  IPv4, using raw IP packets), ipv4icmp (ICMP over IPv4, using raw IP packets),
              ipv4tcp (TCP over IPv4, using raw IP packets), ipv6icmp (ICMPv6  over  IPv6,  using
              raw  IP  packets),  and  ipv6udp  (UDP  over IPv6).  The default protocol is either
              ipv4udp or ipv6udp, as appropriate to the network-layer address associated with the
              hostname  provided.   Compared  with  ipv4udp, the implementation of ipv4raw offers
              finer control over the contents of packet fields, but is otherwise identical.  Note
              that  the  ipv6icmp  and  ipv6udp  options  are  only available if IPv6 support was
              compiled into pcnetm, which can be selected at configure time.  Finally, the ipv4tcp
              option  requires  that  pcap(3)  support be specified at configure time and enabled
              with the -C option.

       -P port
              Select starting UDP port number (the default is  32768).   pcnetm  uses  consecutive
              port  numbers  starting  from this value, counting up.  Care should be taken not to
              use port numbers that are actually in use by network services.

       -q     Quiet mode, suppressing all output.  Useful if writing statistics to  standard  out
              (see the -w option).

       -r file
              Read  measurements  in  from  a file named file, as written by the -w option.  This
              option is useful for experimenting with different analysis algorithms over a  fixed
              data set.

       -R reps
              Set the number of repetitions of each probe packet size to be sent.  The default is
              32 packets of each size.  Smaller values  speed  up  testing,  at  the  expense  of
              accuracy.

       -s hop Set  the  starting  hop  at  which  to begin probing.  The default is 1, so network
              probing will begin at the host adjacent to the  host  where  pcnetm  is  being  run.
              Larger values allow probing to begin farther out from the testing host; this can be
              helpful when attempting to probe outside a local internetwork whose  characterisics
              are well-known.

       -S     Do  SNMP  queries at each hop to determine each router's idea of what it thinks its
              next-hop interface characteristics are.  Use of this features requires the UCD SNMP
              library, as well as enabling at configure-time using --with-snmp.

       -t timeout
              Set  the amount of time (in seconds) that pcnetm will wait for an ICMP error message
              before declaring a packet loss.  The default is 3 seconds.

       -T tos Set the IP Type Of Service bits  for  outgoing  UDP  packets.   This  option  isn't
              terribly  useful  for  a lot of people, but it can be used, for example, to force a
              particular DiffServ codepoint within networks that support this functionality.  For
              values  of  -p  that  use  IPv6  as  a network-layer protocol, this option sets the
              traffic class field in the IPv6 header according to RFC 2460.

       -v     Verbose mode.  While each probe is in progress, print a synopsis of the hop number,
              repetition, and probe packet size on standard out.  Verbose mode mimicks the output
              of pathchar.

       -V     Print version and copyright information and exit.

       -w file
              Write statistics to a datafile named file.  This  file  can  be  read  back  in  by
              specifying  the  -r  option  in a subsequent run of pcnetm for off-line analysis, or
              parsed by other programs for plotting, etc.

              If file is given as  - , then the statistics are written to standard out.  In  this
              case,  the  quiet  flag  -q  may be useful, to avoid cluttering the standard output
              stream.
```

### SEE ALSO

```
   pcap(3), ping(8), traceroute(8), pathchar(8)
```

### Notes

```
       Because pcnetm relies on measurements to drive its estimates of network characteristics, it
       may  occasionally  produce  some  seemingly  odd  results.   Care  should  be  taken  when
       interpreting the output of pcnetm.  For example, the coeffecients of determination for  the
       least squares fit can be useful in seeing how “good” of a fit the bandwidth and round-trip
       time parameters describe the performance seen by the probe packets.   The  coefficient  of
       determination  takes  values  from 0 to 1, where a value of 1 indicates that the estimated
       parameters perfectly fit the data.

       pcnetm was originally named pc, which was either an abbreviation for “path characteristics”
       or “pathchar clone”.
```

### Bugs

```
       Pathchar  automatically  determines  an appropriate maximum packet size to use, based on a
       Path MTU discovery algorithm.  pcnetm relies on the user specifying the maximum packet size
       manually.

       Some  versions  of  Solaris  rate-limit the generation of ICMP error messages.  Any run of
       pcnetm through, or to, a Solaris machine may show abnormally high packet loss rates.   This
       feature  of  Solaris affects traceroute and pathchar as well, but not ping.  Some versions
       of Linux appear to have similar rate-limiting.  In situations such as  this,  the  use  of
       ICMP-based  probes  (selected  by  the -p option) may yield more satisfactory (or at least
       faster) results.

       Timestamps printed after each run are printed relative to the local time zone.  Timestamps
       saved in trace files are expressed as seconds past the epoch.

       There are way too many command-line options
```

### Author

```
       Bruce  A. Mah <bmah@acm.org>.  The author of the original pathchar utility is Van Jacobson
       <van@ee.lbl.gov>.  The algorithms used by pcnetm were coded from Van Jacobson's  viewgraphs
       describing the operation of pathchar.
```

### To Obtain and Build

```
git clone https://github.com/cppnetwork-tools/pcnetm

% ./configure
% make

To enable IPv6 support, give the --with-ipv6 option to configure.  If
there is a directory for IPv6-specific libraries, it can be specified
via an argument to the --with-ipv6 option, for example:

% ./configure --with-ipv6=/usr/local/v6
% make
```

### Contributing

```
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.
Please make sure to update tests as appropriate.
```

pcnetm/CHANGES
PCHAR:  A TOOL FOR MEASURING NETWORK PATH CHARACTERISTICS
Bruce A. Mah
<bmah@kitchenlab.org>
$Id: CHANGES 1086 2005-02-12 20:31:40Z bmah $
---------------------------------------------------------
	
This file lists changes to pchar, in reverse chronological order
(i.e. most recent entries at the top of the file).

pchar-1.5:  Compile fixes, minor bug fixes.  Last planned release of pchar.
(12 February 2005)

	Source code moved from CVS to Subversion.  Although generally
	not user-visible, the version control strings now use
	Subversion's format.

	Builds on recent (RH 9.0-vintage) Linux systems seem to work
	now.

	Some off-by-one errors have been fixed, based on
	patches submitted by Anil Madhavapeddy <anil@recoil.org>.

	pchar with SNMP enabled now has at least a chance of working
	correctly.

	pchar no longer incorrectly aborts after a hop with 100%
	packet loss.

	pchar now does a better job of linking libraries using
	--with-snmp on machines where libsnmp depends on libcrypto,
	thanks to Matt Zimmerman <mdz@debian.org>.

	Some problems using --with-pcap on RedHat 7.0/7.1 (possibly
	other Linux distributions as well) were found thanks to a
	debugging session with Fran Boon <flavour@partyvibe.com>.

pchar-1.4:  Multi-packet probes, TCP probes, kernel-level timestamps.
(13 June 2001)

	pchar now has some limited compatability with the NRL IPv6
	implementation in BSDI 4.1, thanks to some patches originally
	submitted by Antonio Querubin <tony@lava.net>.

	pchar now supports TCP over IPv4 and TCP over IPv6 probes.

	pchar now will use the pcap packet capture library if given
	the --with-pcap flag at configure time (the motivation for
	this is to enable TCP probes, as well as to give potentially
	more accurate timestamps).  It is enabled at
	run-time with the new -C flag.  The selection of which network
	interface to snoop is now controlled by the -i flag (in the
	same way as tcpdump(1)).

	The packet-size increment is now controlled by the -I flag
	(it used to be controlled by -i).

	Add -b burst parameter for measuring store-and-forward
	switched subnets, inspired by LBL's pipechar.  It works for
	most, but not all, protocol types (so far, it is supported by 
	everything except UDP over IPv4; note that raw UDP over IPv4
	supports -b just fine).

	Add a time_t cast to fix building on OpenBSD, submitted by 
	"Angelos D. Keromytis" <angelos@keromytis.org>.

pchar-1.3.2: Solaris and IRIX compatability, a few bugfixes, SUID root
installation as an option.  (9 April 2001)

	Compatability with IRIX has been improved, thanks to 
	Ayamura KIKUCHI <ayamura@ayamura.org>.  A related
	compatability fix was provided by Jeffrey Mogul
	<mogul@pa.dec.com>.

	pchar should now use less memory to hold its results, due to a
	change in the way memory is allocated and managed.

	Henk Langeveld <Henk.Langeveld@holland.sun.com> discovered
	some alignment bugs that were breaking on the Sun compiler.
	These have been fixed.

	pchar now prints the start and end times at the end of each
	run (in local time).

	When reading from tracefiles, the origin address in printed
	output used to be taken from the local machine.  This was a bug.  
	The origin address is now properly taken from the tracefile.

	Handle route changes more gracefully, and keep track of up to
	ten alternative routes for a hop.  Asked for by many folks.

	Providing an invalid argument to the -m argument no longer
	produces a core dump.  Submitted by Pete Wyckoff <pw@osc.edu>.

	pchar now has a compile-time option to build and install
	as SUID root (--with-suid).  It's off by default, and needs to
	be explicitly enabled.

pchar-1.3.1:  Compatability release, as noted below. (29 January 2001)

	Fixed a typo that broke SNMP support.  Reported by "Soren
	S. Jorvang" <soren@wasabisystems.com> and 
	Karl Auerbach <karl@CaveBear.com>.

	Version 5.0 of the Sun C++ compiler defines some slightly
	different preprocessor variables; account for this.  Pointed
	out by Erik Nordmark <Erik.Nordmark@eng.sun.com>.

	The Sun compiler doesn't like arrays whose sizes are
	determined at runtime; this has been fixed.  Pointed out
	by Henk Langeveld <Henk.Langeveld@holland.sun.com>.

	Added a prototype for srandom(3) to support Linux
	compilation on more recent gcc versions.  This requirement
	pointed out by Jaiwant Mulik <jmulik@unix.temple.edu> and
	Karl Auerbach <karl@CaveBear.com>.

pchar-1.3 ("Year of the Snake" release): New analysis types, support
for ICMP and ICMPv6 probe packets, "tiny traceroute" mode, outgoing
interface selection, bugfixes. (24 January 2001)

	Added -M option for operational mode.  This controls whether
	pchar retains its original behavior or collects "trouts"
	(tiny traceroutes).

	Added ICMPv6 over IPv6 probes.

	pchar will no longer try output negative RTT or BW estimates.
	They don't have any physical meaning, and they generate the most
	frequent FAQ.  There were a few more cosmetic changes made to
	the output.

	pchar's analysis methods now take into account the size of the
	response packet when computing bandwidth and delay estimates.
	This makes it possible for to use probes which may return
	non-constant-sized responses (such as the ICMP probes below).

	As a consequence of considering response packet sizes, the UDP
	over IPv6 probes will now return more sane values on the last
	hop (they were too low by about 50%).

	Added ICMP over IPv4 and TCP over IPv4 probes.  (TCP over IPv4
	is not completely functional as of this writing.)

	As a consequence of supporting multiple probe types,
	tracefiles for pchar-1.3 and later are not compatible with
	pchar-1.2 and earlier.  The new tracefiles are smaller than
	the old tracefiles due to less redundant information being
	included.

	Added -l flag to set the source address of outgoing probe
	packets.  If this flag is not specified, attempt to use the
	target address to figure out what the correct source address
	should be.  This is a change in behavior from prior versions,

pcnetm/GetIfInfo.cc
// -*- c++ -*-
// $Id: GetIfInfo.cc 1082 2005-02-12 19:40:04Z bmah $
//

#include <stdio.h>

#ifdef STDC_HEADERS
#include <stdlib.h>
#endif /* STDC_HEADERS */

#ifdef HAVE_UNISTD_H 
#include <unistd.h>
#endif /* HAVE_UNISTD_H */

#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif /* HAVE_STRINGS_H */

#include <errno.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/socket.h>

#include "pc.h"

// UCD SNMP includes
#include "ucd-snmp/ucd-snmp-config.h"
#include "ucd-snmp/asn1.h"
#include "ucd-snmp/snmp_api.h"
#include "ucd-snmp/snmp_impl.h"
#include "ucd-snmp/snmp_client.h"
#include "ucd-snmp/mib.h"
#include "ucd-snmp/snmp.h"
#include "ucd-snmp/system.h"
#include "ucd-snmp/default_store.h"

#include "GetIfInfo.h"
#include "Pctest.h"

static oid *snmp_parse_oid(char *argv, oid *root, size_t *rootlen);

oid * snmp_parse_oid(char *argv,	oid *root, size_t *rootlen)
{
size_t savlen = *rootlen;

if (read_objid(argv,root,rootlen)) {
    return root;
    }

*rootlen = savlen;
if (get_node(argv,root,rootlen)) {
    return root;
    }

return NULL;
}

GetIfInfo::GetIfInfo(void *addr, Pctest *pct) :
			  valid_info(false),
			  Description(NULL),
			  Name(NULL),
			  Contact(NULL),
			  Location(NULL),
			  IfDescription(NULL),
			  IfMtu(0),
			  IfSpeed(0),
			  IfType(0)
{
struct snmp_session session, *ssess;
struct snmp_pdu *pdu;
struct snmp_pdu *response;
struct variable_list *vars;

char *community = "public";

char *intoap;
long ifnumber;

oid name_oid[MAX_OID_LEN];
int status;
size_t name_length;
size_t ndx;
size_t ilen;

char cbuff[128];

typedef enum
  {
  SysDescr_AVBI		= 0,
  SysContact_AVBI	= 1,
  SysName_AVBI		= 2,
  SysLocation_AVBI	= 3,
  IfIndex_AVBI		= 4,
  AVBI_COUNT		= 5
  } query_a_vb_index_e;

char *query_A_names[AVBI_COUNT] =
	{
	"system.sysDescr.0",
	"system.sysContact.0",
	"system.sysName.0",
	"system.sysLocation.0",
    0
	};

typedef enum
  {
  ifDescr_BVBI		= 0,
  ifType_BVBI		= 1,
  ifMtu_BVBI		= 2,
  ifSpeed_BVBI		= 3,
  BVBI_COUNT		= 4
  } query_b_vb_index_e;

char *query_B_names_master[BVBI_COUNT] =
	{
	"interfaces.ifTable.ifEntry.ifDescr.",
	"interfaces.ifTable.ifEntry.ifType.",
	"interfaces.ifTable.ifEntry.ifMtu.",
	"interfaces.ifTable.ifEntry.ifSpeed."
	};

char *query_B_names[BVBI_COUNT];

/* initialize session to default values */
snmp_sess_init(&session);

/* read in MIB database and initialize the snmp library*/
init_snmp("snmpapp");

if (session.version == SNMP_DEFAULT_VERSION)
  {
  session.version = ds_get_int(DS_LIBRARY_ID, DS_LIB_SNMPVERSION);
  }

/* make our engineID something other than what the localhost might
 * be using, otherwise the automatic v3 time-synchronization won't work */
//setup_engineID(NULL, "a bogus text string");

 intoap = pct->GetPrintableAddress(addr);
session.peername = new char[strlen(intoap)+1];
if (session.peername == NULL)
  {
  SOCK_CLEANUP;
  }
strcpy(session.peername, intoap);

session.community = (unsigned char *)community;
session.community_len = strlen(community);

strcpy(cbuff, "ip.ipAddrTable.ipAddrEntry.ipAdEntIfIndex.");
strcat(cbuff, pct->GetPrintableAddress(addr));
query_A_names[IfIndex_AVBI] = cbuff;

SOCK_STARTUP;

ssess = snmp_open(&session);
if (ssess == NULL)
   {
   snmp_sess_perror("snmpget", &session);
   SOCK_CLEANUP;
   }

pdu = snmp_pdu_create(SNMP_MSG_GET);
for (ndx = 0; ndx < AVBI_COUNT; ndx++)
  {
  name_length = MAX_OID_LEN;
  if (!snmp_parse_oid(query_A_names[ndx], name_oid, &name_length))
     {
     snmp_close(ssess);
	 SOCK_CLEANUP;
	 }
  snmp_add_null_var(pdu, name_oid, name_length);
  }

response = NULL;
status = snmp_synch_response(ssess, pdu, &response);
if ((status == STAT_SUCCESS) && (response->errstat == SNMP_ERR_NOERROR))
  {
  for (ndx = 0, vars = response->variables; vars; ndx++, vars = vars->next_variable)
	{
    if (ndx == IfIndex_AVBI)
	  {
	  ifnumber = *(vars->val.integer);
	  }
	else {
	     char *nsp;
		 nsp = new char[vars->val_len + 1];
		 strncpy(nsp, (char *)(vars->val.string), vars->val_len);
		 nsp[vars->val_len] = '\0';
	     switch (ndx)
		   {
		   case SysDescr_AVBI:
			    Description = nsp;
				break;
		   case SysContact_AVBI:
			    Contact = nsp;
				break;
		   case SysName_AVBI:
		        Name = nsp;
				break;
		   case SysLocation_AVBI:
		        Location = nsp;
				break;
		   default:
			    delete nsp;
		   }
	     }
	}

  sprintf(cbuff, "%lu", ifnumber);
  ilen = strlen(cbuff);
  for (ndx = 0; ndx < BVBI_COUNT; ndx++)
	{
	query_B_names[ndx] = new char[ilen + strlen(query_B_names_master[ndx]) + 1];
	strcpy(query_B_names[ndx], query_B_names_master[ndx]);
	strcat(query_B_names[ndx], cbuff);
	}

  snmp_free_pdu(response);
  response = NULL;

  pdu = snmp_pdu_create(SNMP_MSG_GET);
  for (ndx = 0; ndx < BVBI_COUNT; ndx++)
	{
	name_length = MAX_OID_LEN;
	if (!snmp_parse_oid(query_B_names[ndx], name_oid, &name_length))
	  {
	  goto cleanup;
	  }
	snmp_add_null_var(pdu, name_oid, name_length);
	}

  status = snmp_synch_response(ssess, pdu, &response);
  if ((status == STAT_SUCCESS) && (response->errstat == SNMP_ERR_NOERROR))
	{
	for (ndx = 0, vars = response->variables; vars; ndx++, vars = vars->next_variable)
	  {
       switch (ndx)
		 {
		 case ifDescr_BVBI:
		      IfDescription = new char[vars->val_len + 1];
			  strncpy(IfDescription, (char *)(vars->val.string), vars->val_len);
			  IfDescription[vars->val_len] = '\0';
			  break;

		 case ifType_BVBI:
		      IfType = *(vars->val.integer);
			  break;

		 case ifMtu_BVBI:
		      IfMtu = *(vars->val.integer);
			  break;

		 case ifSpeed_BVBI:
		      IfSpeed = *(vars->val.integer);
		      break;
		 }
	  }
	}
  valid_info = true;
  }

cleanup:

if (response)
  {
  snmp_free_pdu(response);
  }
snmp_close(ssess);
SOCK_CLEANUP;
}

GetIfInfo::~GetIfInfo()
{
if (Description) delete Description;
if (Name) delete Name;
if (Contact) delete Contact;
if (Location) delete Location;
if (IfDescription) delete IfDescription;
}

const char unknown_str[] = "-unknown-";

const char *
GetIfInfo::GetDescription(void) const
{
if (!valid_info)
  {
  return unknown_str;
  }
return Description;
}

const char *
GetIfInfo::GetName(void) const
{
if (!valid_info)
  {
  return unknown_str;
  }
return Name;
}

const char *
GetIfInfo::GetContact(void) const
{
if (!valid_info)
  {
  return unknown_str;
  }
return Contact;
}

const char *GetIfInfo::GetLocation(void) const
{
if (!valid_info)
  {
  return unknown_str;
  }
return Location;
}

const char *
GetIfInfo::GetIfDescription(void) const
{
if (!valid_info)
  {
  return unknown_str;
  }
return IfDescription;
}

uint32_t
GetIfInfo::GetIfMtu(void) const
{
if (!valid_info)
  {
  return 0;
  }
return IfMtu;
}

uint32_t
GetIfInfo::GetIfSpeed(void) const
{
if (!valid_info)
  {
  return 0;
  }
return IfSpeed;
}

uint32_t
GetIfInfo::GetIfType(void) const
{
if (!valid_info)
  {
  return 0;
  }
return IfType;
}

const char *iftype_strs[] =
	{
    "other",				//	=  1,	//-- none of the following
	"regular1822",			//	=  2,
	"hdh1822",				//	=  3,
	"ddn-x25",				//	=  4,
	"rfc877-x25",			//	=  5,
	"ethernet-csmacd",		//	=  6,
	"iso88023-csmacd",		//	=  7,
	"iso88024-tokenBus",	//	=  8,
	"iso88025-tokenRing",	//	=  9,
	"iso88026-man",			//	=  10,
	"starLan",				//	=  11,
	"proteon-10Mbit",		//	=  12,
	"proteon-80Mbit",		//	=  13,
	"hyperchannel",			//	=  14,
	"fddi",					//	=  15,
	"lapb",					//	=  16,
	"sdlc",					//	=  17,
	"ds1",					//	=  18,  // -- T-1
	"e1",					//	=  19,  // -- european equiv. of T-1
	"basicISDN",			//	=  20,
	"primaryISDN",			//	=  21,  // -- proprietary serial
	"propPointToPointSerial",//	=  22,
	"ppp",					//	=  23,
	"softwareLoopback",		//	=  24,
	"eon",					//	=  25,  // -- CLNP over IP [11]
	"ethernet-3Mbit",		//	=  26,
	"nsip",					//	=  27,  // -- XNS over IP
	"slip",					//	=  28,  // -- generic SLIP
	"ultra",				//	=  29,  // -- ULTRA technologies
	"ds3",					//	=  30,  // -- T-3
	"sip"					//	=  31   // -- SMDS
	};

const size_t num_iftype_strs = sizeof(iftype_strs)/sizeof(char *);

const char *
GetIfInfo::GetIfTypeString(void) const
{
if (!valid_info)
  {
  return unknown_str;
  }

if ((IfType == 0 ) || (IfType > num_iftype_strs))
  {
  return unknown_str;
  }
return iftype_strs[IfType-1];
}




